Invoke DeleteItemActions on backup deletion

nrb opened the pull request:
Closed #2400

ashish-amarnath commented on internal/delete/delete_item_action.go:
```diff
@@ -0,0 +1,116 @@
+type resolvedAction struct {
+	velero.DeleteItemAction
+
+	resourceIncludesExcludes  *collections.IncludesExcludes
+	namespaceIncludesExcludes *collections.IncludesExcludes
+	selector                  labels.Selector
```
suggest creating a new struct, with just these three fields, that can be used in all ItemAction plugins.
That should save us a lot of code duplication across the plugin types, and also perhaps the hooks.

ashish-amarnath replied:
This will avoid writing `resolveActions` for every plugin.

ashish-amarnath replied:
we can introduce the new type now and use it in the DeleteItemAction plugin and update the other plugin types as a follow-up. WDYT?

nrb replied:
Yeah, I want to unify all of them.

The problem I ran into is that each plugin has a slightly different signature, and this point was missed in the design phase.

There's a lot of duplication between restoration and deletion that I'd like to reduce, though deletion is much simpler as it turns out.

ashish-amarnath commented on pkg/controller/backup_deletion_controller.go:
```diff
@@ -289,6 +294,31 @@ func (c *backupDeletionController) processRequest(req *velerov1api.DeleteBackupR
+		Actions:         actions,
+		Log:             c.logger,
+		DiscoveryHelper: c.helper,
+		Filesystem:      nil,
```
should this be nil? or is this a `TODO`?

nrb replied:
Actually I think it should be populated; I populated it elsewhere in the code right now, but I'll fix that up.

nrb commented:
@ashish-amarnath @carlisia This PR is still in draft because I have a few minor changes I want to make, and I'll have to remove the test implementation (I'll likely just copy it over to the example plugins).

Having the test implementation was helpful because I found some bugs where it wasn't actually getting invoked, but I think as it stands the code is pretty close to what will finally be reviewed, I'm just going to move a couple things around, like the creation of the filesystem and seeing about adding some tests.

nrb commented:
Removed the test plugin from this codebase (will get the PR into the examples repo), and added a test suite based on the restore's `RestoreItemAction` checking.

nrb commented:
Investigating the CI failure.

nrb commented:
Still not sure why CI's failing, as `make ci` passes locally. I'll revisit in the morning.

ashish-amarnath commented:
CI is passing now 🎉

ashish-amarnath commented on internal/delete/delete_item_action_handler_test.go:
```diff
@@ -0,0 +1,264 @@
+	}
+}
+
+func (h *harness) AddItems(t *testing.T, resource *test.APIResource) {
```
nit: Cleaner interface for this could be
```
func (h *harness) AddItems(t *testing.T, resource *[]test.APIResource)
```

ashish-amarnath replied:
Also in https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore_test.go#L2821 this is not an exported method. Any reason to change that here?

ashish-amarnath replied:
This is another opportunity to dedup code.
Suggest adding `TODOs` for us to do code dedup and cleanup as a follow-up.

ashish-amarnath commented on pkg/restore/restore.go:
```diff
@@ -535,6 +504,7 @@ func getNamespace(logger logrus.FieldLogger, path, remappedName string) *v1.Name
 	}
 }
 
+// TODO: this should be combined with DeleteItemActions at some point.
```
👍

nrb commented on internal/delete/delete_item_action_handler.go:
```diff
@@ -0,0 +1,218 @@
+					// Since we want to keep looping even on errors, log them instead of just returning.
+					if err != nil {
+						itemLog.WithError(err).Error("plugin error")
+
```
Yeah, it would. Let me do some testing/checking, because I _think_ the error returned will actually have the plugin's information passed back through gRPC, but I don't remember. I don't know if the name's directly available here.

nrb commented on internal/delete/delete_item_action_handler.go:
```diff
@@ -0,0 +1,218 @@
+
+	// No actions installed and no error means we don't have to continue;
+	// just do the backup deletion without worrying about plugins.
+	if (ctx.resolvedActions == nil || len(ctx.resolvedActions) == 0) && err == nil {
```
I had that here, but it was failing tests. I'll try it again.

nrb commented on internal/delete/delete_item_action_handler_test.go:
```diff
@@ -0,0 +1,264 @@
+				h.AddItems(t, r)
+			}
+
+			actions := []velero.DeleteItemAction{}
+			for action := range tc.actions {
+				actions = append(actions, action)
+			}
```
Actually, it can't because of a type mismatch.

`./delete_item_action_handler_test.go:161:5: cannot use tc.actions (type map[*recordResourcesAction][]string) as type []velero.DeleteItemAction in field value`

nrb replied:
Adding some comments to make this a little more clear.

nrb commented on internal/delete/delete_item_action_handler.go:
```diff
@@ -0,0 +1,218 @@
+}
+
+// unmarshal reads a JSON file into an Unstructured data object.
+func (ctx *Context) unmarshal(filePath string) (*unstructured.Unstructured, error) {
```
Moved to `archive.Unmarshal` with that interface.

nrb commented:
@ashish-amarnath I think I've addressed all your comments now, though not exactly. Some are going to be deferred for later clean up.

@carlisia Please take a look!

ashish-amarnath commented on internal/delete/delete_item_action_handler.go:
```diff
@@ -160,10 +160,10 @@ func (ctx *Context) getApplicableActions(groupResource schema.GroupResource, nam
 
 // unmarshal reads a JSON file into an Unstructured data object.
-func (ctx *Context) unmarshal(filePath string) (*unstructured.Unstructured, error) {
+func unmarshal(fs filesystem.Interface, filePath string) (*unstructured.Unstructured, error) {
```
Now that we have `archive.Unmarshal` this can be removed?

ashish-amarnath replied:
I forgot I made this comment. So, LGTM pending this change 😄

nrb replied:
Thought I removed it, but had missed it. Updated!

ashish-amarnath reviewed (dismissed):
LGTM! 🚀
